### PR TITLE
Improve programs/cert_write with a way to set extended key usages

### PR DESCRIPTION
**Pull request by NWilson**
I find the `cert_write` tool is much easier to use that `openssl ca`, and it supports almost all the features I need to create test certificates. It would make generating test cases much easier though if I could use it to set extended key usages.

This PR adds a function, `mbedtls_x509write_crt_set_ext_key_usage`, and an option in `cert_write` to set some usages on the new certificate.

### CODE REVIEW

**Comment by NWilson** [comment]
Thanks Manuel for those comments, the mbedTLS style is hard to pick up! I've tweaked those changes, and pushed another commit that lets you set the signature digest as well.

**Comment by mpg** [comment]
Thanks! One more thing: we'll need at least one test case for the new function you've added. Is that something you would be willing to contribute too? If not, no worries, we'll write it ourselves. It might just delay a bit our merging of your patch.

File: programs/x509/cert_write.c [hunk 1]
```diff
@@ -303,6 +322,17 @@ int main( int argc, char *argv[] )
             if( opt.max_pathlen < -1 || opt.max_pathlen > 127 )
                 goto usage;
         }
+        else if( strcmp( p, "sig_alg") == 0 )
+        {
+            if( strcmp( q, "SHA-1" ) == 0 )
+                opt.sig_alg = MBEDTLS_MD_SHA1;
+            else if( strcmp( q, "SHA-245" ) == 0 )
```
Comment (NWilson) [inline programs/x509/cert_write.c]: Oops, I'll fix that!

**Comment by NWilson** [comment]
Thanks, I've tried getting my head around your test framework before, it's a bit custom, so I think it would take you less time! I could make a test case, but not until I get back from holiday in a week.

I'll make some time and add the test around October if you haven't been able to help me out and do it by then.

Just out of interest, how do you generate test certs yourself? You don't have any code in mbedTLS for generating extensions, so you don't use OpenSSL do you?

**Comment by mpg** [comment]
Currently we use OpenSSL or GnuTLS to generate most test certs (some come from external sources), in order to ensure interoperability too.

Agreed about our tests framework: it takes some getting used to, and is not much documented. Your plan sounds good: we'll probably do it, and if it turns out we're too much pressed by other tasks, you can always have a go at it.

**Comment by ciarmcom** [comment]
Automatic CI verification build not done, please verify manually.

**Comment by NWilson** [comment]
Hi @mpg, I've just added the test function. The certificates we generate parse fine using OpenSSL, so I checked them in to the data_files directory, just to catch regressions.

**Comment by mpg** [comment]
Sounds good! I'll have a look when I get the occasion. (Obviously you know we have at least one other issue that is more pressing than this one...)

**Comment by NWilson** [comment]
Hi Manuel, I've got three open PRs at the moment, it's been a month since any of them have been touched. Could I ask whether there's any progress on merging them, or giving feedback if there's further work needed on them?

I don't want to give you a big backlog of PRs, so I'm waiting until the current batch are accepted before submitting some more branches I've got ready.

**Comment by sbutcher-arm** [comment]
We're very conscious that we have a large backlog at the moment which we haven't had time to clear down. We're currently going through planning for our next feature release and I hope to resolve the bulk of these then, so expect a response in the New Year.

Apologies for the delay.

**Comment by NWilson** [comment]
Thanks, I'm not in a rush so I'll gently send another ping in a month's time.

**Comment by sbutcher-arm** [comment]
We appreciate the feedback, so feel free to nudge us again in the future!

**Comment by billybednar** [comment]
I noticed this PR marks the extension as critical. I'm not sure about the implications but every certificate I've seen has it as non-critical.

**Comment by yanesca** [comment]
Thank you very much for your remark! We will look into it when processing this PR.

**Comment by josesimoes** [comment]
👍🏻  for having this reviewed and merged! Please! 😁

**Comment by daverodgman** [comment]
In historical review, we agreed that given the age of this PR, we'll adopt and update / fix conflicts / progress it (happy for @NWilson to update if preferred - please shout if you prefer to do so).

**Comment by daverodgman** [comment]
Updated in #6121 - closing this version.

**Pull request closed without merging**